### PR TITLE
adding custom filters to FilterOrderRegistration  

### PR DESCRIPTION
**Pull request by theexiile1305**
This pull request let custom filters to be added to the FilterOrderRegistration in order to solve the issue mentioned in #9787.

Closes gh-9787

### CODE REVIEW

**Comment by Stexxen** [comment]
Hi @theexiile1305, I don't work for pivotal and personally I want to see this issue resolved asap, but I suspect they are going to want a couple of tests as well?
Just my 2 cents...

**Comment by theexiile1305** [comment]
Hi @Stexxen, thank you for the reply. Yea, I give it a try to add some tests 😄

File: config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java [hunk 1]
```diff
@@ -2653,6 +2653,7 @@ public HttpSecurity addFilterBefore(Filter filter, Class<? extends Filter> befor
 	private HttpSecurity addFilterAtOffsetOf(Filter filter, int offset, Class<? extends Filter> registeredFilter) {
 		int order = this.filterOrders.getOrder(registeredFilter) + offset;
 		this.filters.add(new OrderedFilter(filter, order));
+		this.filterOrders.add(filter.getClass());
```
Comment (gesellix) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java:2656]: This fix should be added to the `[#addFilter(Filter)](https://github.com/spring-projects/spring-security/pull/9832/files#diff-582061af3df753b11cdf7e63b3c2090816c59ef44480bf0c46523354fe3b5fc5L2666)` method just below this one, shouldn't it?

Comment (theexiile1305) [reply config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java]: Yea, you are aright. I've changed the line like you suggested.

Comment (gesellix) [reply config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java]: Oh wait, I thought that both methods need the fix 😅

Comment (theexiile1305) [reply config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java]: Sorry, for that misconception 😅 .

**Comment by theexiile1305** [comment]
@rwinch Hey there 😃 
Should I add some unit or integration tests or both to this pull request? What do you prefer? Do you think, that the changes in this pull request are right so far? 
Thank you  for the feedback.

**Review by rwinch — CHANGES_REQUESTED** [review]
Thanks for the PR! I've provided feedback inline.

You can add tests to [`HttpSecurityAddFilterTest`](https://github.com/spring-projects/spring-security/blob/20751d206332d1dd0900ae4bc1fe103ccf6b43db/config/src/test/java/org/springframework/security/config/annotation/web/builders/HttpSecurityAddFilterTest.java#L44). Please be sure to include a test for the original issue and to cover the scenarios I mentioned inline.

**Comment by marcusdacoregio** [comment]
I am going to work on the PR and clean it up based on @rwinch suggestions.

**Comment by ch4mpy** [comment]
I created this PR https://github.com/spring-projects/spring-security/pull/9898 which was closed for being a "duplicate".

It contains unit tests, solves the bug and code modification is really small.

Feel free to copy / paste from there if it helps to get this solved quickly, I don't care about having credits but do care this bug is solved as it is a show stopper for me to migrate to spring-boot 2.5

**Comment by jgrandja** [comment]
@theexiile1305 We haven't received any updates based on @rwinch feedback so we went ahead with the changes.

I'll close this in favour of  gh-9902.

**Pull request closed without merging**